File: 238. Product of Array Except Self.py
# 238. Product of Array Except Self.      -Medium-

# Topic: Array, Prefix Sum.

"""
### Task:
---
Given an integer array nums, return an array answer such that answer[i] is equal to the product of 
all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.

Example 1:
Input: nums = [1,2,3,4]
Output: [24,12,8,6]

Example 2:
Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]

Constraints:
2 <= nums.length <= 10^5
-30 <= nums[i] <= 30
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 
Follow up: Can you solve the problem in O(1) extra space complexity? (The output array does not count 
           as extra space for space complexity analysis.)

### Testcase:
---
[1,2,3,4]
[-1,1,0,-3,3]

### Code:
---
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
"""
# Solution: -------------------------------------

class Solution:
    def productExceptSelf(self, nums):
        n = len(nums)
        # Initialize two arrays: 'left' to store the product of elements to the left of each index,
        # and 'result' to store the final product values.
        left, result = [0] * n, [0] * n
        
        # The product of elements to the left of the first element is 1 (since there are no elements to the left)
        left[0] = 1
        for i in range(1, n):
            # left[i] is the product of all elements to the left of 'nums[i]', which is
            # the product up to 'nums[i - 1]' (stored in left[i - 1]) times 'nums[i - 1]'
            left[i] = nums[i - 1] * left[i - 1]
        
        # 'right' will keep track of the product of elements to the right of the current index
        right = 1
        for i in range(n - 1, -1, -1): # Loop through the array in reverse
            # For each index i, multiply the product of all elements to the left (left[i])
            # with the product of all elements to the right (right) to get the final product
            result[i] = left[i] * right
            # Update 'right' to include the current element, for the next iteration
            right *= nums[i]
        
        return result


# Test the function with the provided test cases
sol = Solution()
test_case_1 = [1, 2, 3, 4]
test_case_2 = [-1, 1, 0, -3, 3]

output_1 = sol.productExceptSelf(test_case_1)
output_2 = sol.productExceptSelf(test_case_2)

output_1, output_2



# Description: ==================================
'''
In this code:
- The `left` array is used to store the cumulative product of all elements to the left of each index in the input array.
- The `right` variable is used to keep track of the cumulative product of elements to the right of the current index as we iterate 
  backward through the array.
- The final result is obtained by multiplying the corresponding values from the `left` array and the `right` variable.


The "Product of Array Except Self" problem requires us to calculate, for each element in the array, the product of all other 
elements in the array without including the current element. The challenge is to do this efficiently, in O(n) time complexity, 
and without using division.

### Algorithm:
1. Create two arrays, `left` and `right`, each of the same length as the input array `nums`. 
2. Fill the `left` array such that `left[i]` contains the product of all elements to the left of `i` in `nums`.
3. Similarly, fill the `right` array such that `right[i]` contains the product of all elements to the right of `i` in `nums`.
4. Create an output array `result` of the same length as `nums`.
5. For each index `i` in `nums`, calculate `result[i]` as `left[i] * right[i]`.
6. Return `result`.

### Optimization for Space Complexity:
To optimize space complexity, we can avoid using the `right` array. Instead, we can directly accumulate the product from the 
right in a variable and use it to update the `result` array.


### Explanation:
- `left` array stores the cumulative product of elements to the left of the current index.
- We iterate from right to left, keeping track of the cumulative product (`right`) and multiplying it with the corresponding 
 `left` value to get the final product for each index.
- This approach takes O(n) time and O(1) space (excluding the output array).

### Example Execution:
- Input: `[1,2,3,4]`
- `left` array after first loop: `[

The implemented solution correctly solves the "Product of Array Except Self" problem for the given test cases:

1. For the input `[1, 2, 3, 4]`, the output is `[24, 12, 8, 6]`. 
   - Here, each element in the output array is the product of all other elements in the input array, except for the element at the 
    corresponding position.

2. For the input `[-1, 1, 0, -3, 3]`, the output is `[0, 0, 9, 0, 0]`.
   - Notably, when there's a `0` in the input array (like in this case), the product for all positions except where the zero is 
     located will be `0`, since multiplying by zero results in zero. The position corresponding to the zero in the input array 
     gets the product of all other (non-zero) elements.

This solution meets the requirements of the problem:
- It runs in O(n) time complexity, iterating over the array twice.
- It doesn't use the division operation.
- It has O(1) extra space complexity, excluding the output array, as it uses only a constant amount of additional space for the 
  `left` and `right` variables.

'''

File: 238. Product of Array Except Self.rs
// 238. Product of Array Except Self.      -Medium-

// Topic: Array, Prefix Sum.

/*
### Task:
---
Given an integer array nums, return an array answer such that answer[i] is equal to the product of 
all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.

Example 1:
Input: nums = [1,2,3,4]
Output: [24,12,8,6]

Example 2:
Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]

Constraints:
2 <= nums.length <= 10^5
-30 <= nums[i] <= 30
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 
Follow up: Can you solve the problem in O(1) extra space complexity? (The output array does not count 
           as extra space for space complexity analysis.)

### Testcase:
---
[1,2,3,4]
[-1,1,0,-3,3]

### Code:
---
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        
    }
}
*/
//Solution: -------------------------------------

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len(); // Get the length of the input array
        let mut answer = vec![1; n]; // Initialize the answer array with 1s. This array will eventually contain the product of all elements except the one at the current index.

        // Calculate left products for each element
        let mut left_product = 1; // Initialize left_product which will hold the product of all elements to the left of the current index
        for i in 0..n {
            // For each element in the array:
            answer[i] = left_product; // Set the current index of answer to left_product. Initially, for the first element, it will be 1 (as there are no elements to the left).
            left_product *= nums[i]; // Update left_product by multiplying it with the current element. This prepares left_product for the next index in the array.
        }

        // Calculate right products for each element and finalize the answer
        let mut right_product = 1; // Initialize right_product which will hold the product of all elements to the right of the current index
        for i in (0..n).rev() {
            // Traverse the array in reverse order:
            answer[i] *= right_product; // Multiply the current index of answer with right_product. This combines the product of all elements to the left and right of the current index.
            right_product *= nums[i]; // Update right_product by multiplying it with the current element. This prepares right_product for the next index (to the left) in the array.
        }

        answer // Return the final answer array
    }
}


// Description: ---------------------------------
/*
To solve the problem presented, we'll implement the `product_except_self` function in Rust. The function must compute the product of all elements in the `nums` array except the current element, without using division and ensuring a time complexity of O(n).

The key strategy here is to use two arrays (or one array with two passes if optimizing for space): 
1. `left_products`: At each index `i`, it stores the product of all elements to the left of `i`.
2. `right_products`: At each index `i`, it stores the product of all elements to the right of `i`.

Finally, we calculate the output array where each element is the product of its corresponding values in `left_products` and `right_products`.

Explanation:
- We initialize `answer` as a vector of 1's, since the product of zero numbers is 1.
- We first traverse the array from left to right, updating `left_product` with the cumulative product at each step, and setting `answer[i]` to `left_product`.
- Then, we traverse from right to left. We update `right_product` with the cumulative product and multiply `answer[i]` by `right_product`. This step combines the left and right products for each element.

This solution has a time complexity of O(n) as it traverses the array twice, and a space complexity of O(1) (not counting the output array, as per the problem's follow-up requirement).

*/